src/methods/call.js

```diff
@@ -326,60 +327,114 @@ export function OrdinaryCallEvaluateBody(
       // 4. Return Completion{[[Type]]: return, [[Value]]: G, [[Target]]: empty}.
       return new ReturnCompletion(G, realm.currentLocation);
     } else {
-      // 1. Perform ? FunctionDeclarationInstantiation(F, argumentsList).
-      Functions.FunctionDeclarationInstantiation(realm, F, argumentsList);
-
-      // 2. Return the result of EvaluateBody of the parsed code that is the value of F's
-      //    [[ECMAScriptCode]] internal slot passing F as the argument.
-      let priorSavedCompletion = realm.savedCompletion;
+      if (!realm.useAbstractInterpretation || realm.pathConditions.length === 0) return normalCall();
+      let savedIsSelfRecursive = F.isSelfRecursive;
       try {
-        realm.savedCompletion = undefined;
+        F.isSelfRecursive = false;
+        let effects = realm.evaluateForEffects(guardedCall);
+        if (F.isSelfRecursive) {
+          AbstractValue.reportIntrospectionError(F, "call to function that calls itself");
```

Comment:
nit: maybe the word recursion should be in the error message.

Reply:
This error message is going to be very short lived.